include/quic/network.hpp

```diff
@@ -12,172 +12,96 @@ extern "C"
 #include <thread>
 #include <uvw.hpp>
 
-#include "client.hpp"
 #include "context.hpp"
 #include "crypto.hpp"
-#include "handler.hpp"
-#include "server.hpp"
 #include "utils.hpp"
 
+using oxen::log::slns::source_location;
+
 namespace oxen::quic
 {
+    template <typename... T>
+    void loop_trace_log(
+            const log::logger_ptr& cat_logger,
+            [[maybe_unused]] const source_location& location,
+            [[maybe_unused]] fmt::format_string<T...> fmt,
+            [[maybe_unused]] T&&... args)
+    {
+#if defined(NDEBUG) && !defined(OXEN_LOGGING_RELEASE_TRACE)
+        // Using [[maybe_unused]] on the *first* ctor argument breaks gcc 8/9
+        (void)cat_logger;
+#else
+        if (cat_logger)
+            cat_logger->log(log::detail::spdlog_sloc(location), log::Level::trace, fmt, std::forward<T>(args)...);
+#endif
+    }
+
+	class Endpoint;
+
     class Network
     {
-        friend class Handler;
+        using Job = std::pair<std::function<void()>, source_location>;
+        using handle_address_pair = std::pair<const Address, std::shared_ptr<uv_udp_t>>;
 
       public:
-        Network(std::shared_ptr<uvw::loop> loop_ptr, std::thread::id loop_thread_id);
+
+        Network(std::shared_ptr<uvw::loop> loop_ptr, std::thread::id thread_id);
         Network();
         ~Network();
 
+		std::shared_ptr<Endpoint> endpoint(Address& local_addr);
+        std::shared_ptr<Endpoint> endpoint(const Address& local_addr);
+
+        void close();
+
+		// Find and return the endpoint with the given local address; returns nullptr if not found
+		Endpoint* get_endpoint(const Address& local);
```

Comment:
should probably return `shared_ptr` as the functions above do.

Reply:
It's unnecessary now, so deleting it entirely. I changed `udp_data` to store an `Endpoint&` rather a shared_ptr to the handler which eliminated the call to `get_endpoint`